src/libsystemd-network/sd-dhcp6-lease.c

```diff
@@ -587,6 +600,67 @@ int sd_dhcp6_lease_get_captive_portal(sd_dhcp6_lease *lease, const char **ret) {
         return 0;
 }
 
+int dhcp6_lease_add_vendor_option(sd_dhcp6_lease *lease, const uint8_t *optval, size_t optlen, size_t base_offset) {
+        int r;
+        uint16_t enterprise_id;
+
+        assert(lease);
+        assert(optval || optlen == 0);
+
+        lease->vendor_options = set_new(NULL);
+        enterprise_id = unaligned_read_be16(optval + offsetof(DHCP6Option, len));
+
+        for (size_t offset = 0; offset < optlen;) {
+                const uint8_t *subval;
+                size_t sublen;
+                uint16_t subopt;
+
+                r = dhcp6_parse_vendor_option(optval, optlen, &offset, &subopt, &sublen, &subval);
+                if (r < 0)
+                        return r;
+
+                switch (subopt) {
+                        case DHCP6_VENDOR_SUBOPTION_BASE ... DHCP6_VENDOR_SUBOPTION_LAST:
+                                r = dhcp6_lease_insert_vendor_option(lease, subopt, subval, sublen, enterprise_id);
+                                if (r < 0)
+                                        return r;
+
+                        break;
+                }
+        }
+        return 0;
+}
+
+int dhcp6_lease_insert_vendor_option(sd_dhcp6_lease *lease, uint16_t option_code, const void *data, size_t len, uint16_t enterprise_id) {
+        _cleanup_(dhcp6_vendor_option_freep) sd_dhcp6_option *option = NULL;
+        int r;
+
+        assert(lease);
+
+        option = new(sd_dhcp6_option, 1);
+        if (!option)
+                return -ENOMEM;
+
+        *option = (sd_dhcp6_option) {
```

Comment:
Please also set `.n_ref = 1`, and replace the cleanup function with `sd_dhcp6_option_unrefp`.

Reply:
Fixed